go/test/endtoend/recovery/shardedpitr/shardedpitr_test.go

```diff
@@ -0,0 +1,621 @@
+// - run InitShardMaster
+// - insert some data
+// - Take a master backup
+// - Point binlog server to master
+// - insert some more data
+// - Create restore snapshot
+// - Insert some more data
+// - perform a resharding
+// - Create a recover keyspace which recovers data till the snapshot time
+// - Point binlog server to master of  both shards
+// - take a backup of both new shards
+// - insert more data on the masters of both shards
+// - Create snapshot time
+// - create a recovery keyspace which recovers data from both the shards till the snpashot time
+func TestShardedPITRRecovery(t *testing.T) {
+	defer cluster.PanicHandler(nil)
+	_, err := initializeCluster(t)
+	defer clusterInstance.Teardown()
+	require.NoError(t, err)
+
+	// Creating the table
+	_, err = master.VttabletProcess.QueryTablet(createTable, keyspaceName, true)
+	require.NoError(t, err)
+
+	insertRow(t, 1, "prd-1", false)
+	insertRow(t, 2, "prd-2", false)
+
+	cluster.VerifyRowsInTabletForTable(t, replica, keyspaceName, 2, "product")
+
+	// backup the replica
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("Backup", replica.Alias)
+	require.NoError(t, err)
+
+	// check that the backup shows up in the listing
+	output, err := clusterInstance.ListBackups("ks/0")
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(output))
+
+	//start the binlog server and point it to master
+	bs, err := newBinlogServer(hostname, clusterInstance.GetAndReservePort())
+	defer bs.stop()
+	require.NoError(t, err)
+
+	err = bs.start(mysqlMaster{
+		hostname: binlogHost,
+		port:     master.MysqlctlProcess.MySQLPort,
+		username: mysqlUserName,
+	})
+	require.NoError(t, err)
+
+	// now insert some more data to simulate the changes after regular backup
+	// every insert has some time lag/difference to simulate the time gap between rows
+	// and when we recover to certain time, this time gap will be able to identify the exact eligible row
+	var timeToRecover string
+	for counter := 3; counter <= 6; counter++ {
+		if counter == 4 { // we want to recovery till this, so noting the time
+			tm := time.Now().Add(1 * time.Second).UTC()
+			timeToRecover = tm.Format(time.RFC3339)
+		}
+		insertRow(t, counter, fmt.Sprintf("prd-%d", counter), true)
+	}
+
+	createRestoreSnapshot(t, timeToRecover, RestoreKSName)
+
+	err = clusterInstance.VtctlclientProcess.ApplyVSchema(keyspaceName, vSchema)
+	require.NoError(t, err)
+
+	// create the split shards
+	shardedTablets := []*cluster.Vttablet{shard0Master, shard0Replica, shard1Master, shard1Replica}
+	for _, tablet := range shardedTablets {
+		// Creating a user so that binlog server can read all the events.
+		mysqlUserName := "vt_dba"
+		queryCmds := []string{
+			fmt.Sprintf("CREATE USER '%s'@'%%';", mysqlUserName),
+			fmt.Sprintf("GRANT ALL ON *.* TO '%s'@'%%';", mysqlUserName),
+			fmt.Sprintf("GRANT GRANT OPTION ON *.* TO '%s'@'%%';", mysqlUserName),
+			fmt.Sprintf("create database %s;", "vt_ks"),
+			"FLUSH PRIVILEGES;",
+		}
+
```

Comment:
Move these to a common place, I see this piece of code is there in `initializeCluster` .

Reply:
Moved the code to initializeCluster func